tools/json2bin/json2bin.py

```diff
@@ -0,0 +1,168 @@
+            data schema.
+        '''
+        binary = bytearray([])
+        for key in self.registry:
+            if key == 'name':
+                self.registry[key](data[key])
+            elif key.startswith('padding_'):
+                size, val = self.registry[key]
+                binary = binary + val.to_bytes(size, 'little')
+            else:
+                parse_func, size, const_type = self.registry[key]
+                data_val = self._walk(data, key.split('.'))
+                binary = binary + parse_func(data_val, size, const_type)
+        
+        return binary
+
+
+def pack_flags(flags: list[str], size: int, consts: type[Flag]) -> bytes:
+    '''
+        Pack a list of flag constants into a bitmask. Flag values are defined
+        by the specified consts type, which must descend from the enum.Flag
+        type.
+    '''
+    result = consts(0)
+    for flag_name in flags:
+        result = result | consts[flag_name]
+
+    return result.value.to_bytes(size, 'little')
+
+
+def parse_const(val: str, size: int, consts: type[Enum]) -> bytes:
+    '''
+        Simple parse wrapper for a value belonging to a set of constants,
+        represented in JSON as a raw string name.
+    '''
+    return consts[val].value.to_bytes(size, 'little')
+
+
+def parse_int(val: int, size: int, _consts: type[Enum]) -> bytes:
+    '''
+        Simple parse wrapper for an integer.
+    '''
+    return val.to_bytes(size, 'little')
+
+
+def parse_sint(val: int, size: int, _consts: type[Enum]) -> bytes:
+    '''
+        Simple parse wrapper for a signed integer.
+    '''
+    if val < 0:
+        val = val + (1 << (size * 8))
+    
+    return val.to_bytes(size, 'little')
+
+
+def parse(fname_in: str, schema: Parser) -> bytes:
+    with open(fname_in, 'r', encoding='utf8') as input_file:
+        input_json = json.load(input_file)
+        return schema.parse(input_json)
+
+
+def write(output_bin: bytes, output_idx: int, output_dir: str | None):
+    output_fname = f'{output_idx:04}.bin'
+    if output_dir:
+        output_fname = os.path.join(output_dir, output_fname)
+    
+    with open(output_fname, 'wb') as output_file:
+        output_file.write(output_bin)
+
+
+def process(fname_in: str,
+            schema: Parser,
+            output_dir: str | None,
+            index_func: FunctionType):
+    if fname_in[-5:] != '.json':
+        return
+    
+    output_bin = parse(fname_in, schema)
+    simple_name = os.path.basename(os.path.normpath(fname_in))
+    output_idx = index_func(simple_name.split(".")[0].upper())
+    write(output_bin, output_idx, output_dir)
+
+
+def json2bin(target: str,
+             schema: Parser,
+             output_dir: str | None,
+             index_func: FunctionType,
+             batch_mode: bool):
+    if not os.path.exists(output_dir):
+        os.makedirs(output_dir)
+    
+    if batch_mode:
+        for fname_in in os.listdir(target):
```

Comment:
Might eventually want to change to a recursive glob in the future.

Reply:
Yeah, it's on my mind. Out of scope for now.

Reply:
I wound up doing this anyways.